core/src/main/scala/kafka/controller/PartitionStateMachineV2.scala

```diff
@@ -0,0 +1,268 @@
+    }
+  }
+
+  private def initializeLeaderAndIsrForPartitions(partitions: Seq[TopicAndPartition]): Unit = {
+    val replicasPerPartition = partitions.map(partition => partition -> controllerContext.partitionReplicaAssignment(partition))
+    val liveReplicasPerPartition = replicasPerPartition.map { case (partition, replicas) =>
+        val liveReplicasForPartition = replicas.filter(replica => controllerContext.isReplicaOnline(replica, partition))
+        partition -> liveReplicasForPartition
+    }
+    val (partitionsWithoutLiveReplicas, partitionsWithLiveReplicas) = liveReplicasPerPartition.partition(kv => kv._2.isEmpty)
+
+    partitionsWithoutLiveReplicas.foreach { case (partition, replicas) =>
+      val failMsg = s"Controller $controllerId epoch ${controller.epoch} encountered error during state change of " +
+        s"partition $partition from New to Online, assigned replicas are " +
+        s"[${replicas.mkString(",")}], live brokers are [${controllerContext.liveBrokerIds}]. No assigned " +
+        "replica is alive."
+      logFailedStateChange(partition, NewPartition, OnlinePartition, new StateChangeFailedException(failMsg))
+    }
+    val leaderIsrAndControllerEpochAndContexts = partitionsWithLiveReplicas.map { case (partition, liveReplicas) =>
+      val leaderAndIsr = LeaderAndIsr(liveReplicas.head, liveReplicas.toList)
+      val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(leaderAndIsr, controller.epoch)
+      partition -> (leaderIsrAndControllerEpoch, (partition, leaderIsrAndControllerEpoch))
+    }
+    val createResponses = KafkaControllerZkUtils.createTopicPartitionStates(zookeeperClient, leaderIsrAndControllerEpochAndContexts.toMap)
+    createResponses.foreach { createResponse =>
+      val code = Code.get(createResponse.rc)
+      val (partition, leaderIsrAndControllerEpoch) = createResponse.ctx.asInstanceOf[(TopicAndPartition, LeaderIsrAndControllerEpoch)]
+      if (code == Code.OK) {
+        controllerContext.partitionLeadershipInfo.put(partition, leaderIsrAndControllerEpoch)
+        controllerBrokerRequestBatch.addLeaderAndIsrRequestForBrokers(leaderIsrAndControllerEpoch.leaderAndIsr.isr,
+          partition.topic,
+          partition.partition,
+          leaderIsrAndControllerEpoch,
+          controllerContext.partitionReplicaAssignment(partition),
+          isNew = true)
+        partitionState.put(partition, OnlinePartition)
+      } else if (code == Code.NODEEXISTS) {
+        val failMsg = s"Controller $controllerId epoch ${controller.epoch} encountered error while changing " +
+          s"partition $partition's state from New to Online since LeaderAndIsr path already exists."
+        logFailedStateChange(partition, NewPartition, OnlinePartition, new StateChangeFailedException(failMsg))
+      }
+    }
+  }
+
+  private def electLeaderForPartitions(partitions: Seq[TopicAndPartition], partitionLeaderElectionStrategy: PartitionLeaderElectionStrategy): Unit = {
```

Comment:
It seems that we should do this in a while loop since we do conditional update to the leaderAndIsr path in ZK and need to retry on BadVersion.